utils/man/man1/rapl-set.1
.TH "RAPL\-SET" "1" "2021-04-14" "powercap" "rapl\-set"
.SH "NAME"
.LP
rapl\-set \- set RAPL configurations
.SH "SYNPOSIS"
.LP
\fBrapl\-set\fP [\fIOPTION\fP]...
.SH "DESCRIPTION"
.LP
This utility is deprecated, use
.BR powercap\-set (1)
instead.
.LP
Sets Intel Running Average Power Limit (RAPL) configurations.
.LP
Note that the \-p and \-z options are different from
.BR powercap\-set (1).
.LP
This software requires an Intel processor (Sandy Bridge or newer), Linux
kernel 3.13 or newer compiled with \fBCONFIG_POWERCAP\fR and
\fBCONFIG_INTEL_RAPL\fR enabled, and the \fBintel_rapl\fR kernel module to
be loaded.
.SH "OPTIONS"
.LP
.TP
\fB\-h,\fR \fB\-\-help\fR
Prints out the help screen
.TP
\fB\-p,\fR \fB\-\-zone\fR=\fIZONE\fP
The zone number (0 by default)
.TP
\fB\-p,\fR \fB\-\-package\fR=\fIPACKAGE\fP
Deprecated, use \-\-zone instead
.TP
\fB\-z,\fR \fB\-\-subzone=\fR\fISUBZONE\fP
The subzone number (none by default)
.TP
\fB\-c,\fR \fB\-\-constraint\fR=\fICONSTRAINT\fP
The constraint number (none by default)
.LP
The following zone-level arguments may be used together (\-z/\-\-subzone is optional):
.TP
\fB\-j,\fR \fB\-\-z\-energy\fR
Reset zone energy counter
.TP
\fB\-e,\fR \fB\-\-z\-enabled=1|0\fR
Enable/disable a zone
.LP
The following constraint-level arguments may be used together and require
\-c/\-\-constraint (\-z/\-\-subzone is optional):
.TP
\fB\-l,\fR \fB\-\-c\-power\-limit=UW\fR
Set constraint power limit
.TP
\fB\-s,\fR \fB\-\-c\-time\-window=US\fR
Set constraint time window
.SH "EXAMPLES"
.LP
Note that \-p/\-\-zone=0 is used by default, allowing for simpler
commands on single-socket systems.
.TP
\fBrapl\-set \-e 1\fP
Enable zone 0 (implicitly), which is usually named \fBpackage\-0\fR.
.TP
\fBrapl\-set \-p 0 \-e 1\fP
Enable zone 0 (explicitly).
.TP
\fBrapl\-set \-p 0 \-z 1 \-e 1\fP
Enable zone 0, subzone 1, which is usually the \fBuncore\fR or
\fBdram\fR subzone of \fBpackage\-0\fR, depending on the system.
.TP
\fBrapl\-set \-p 0 \-c 1 \-l 25000000\fP
Set a power cap of 25 Watts (25000000 uW) on zone 0, constraint 1,
which is usually the \fBshort_term\fR constraint for \fBpackage\-0\fR.
.TP
\fBrapl\-set \-p 1 \-z 0 \-c 0 \-l 15000000 \-s 976\fP
Set a power cap of 15 Watts (15000000 uW) and time window of 976
microseconds on zone 1, subzone 0, constraint 0, which is usually the
\fBlong_term\fR (and only) constraint for the \fBcore\fR subzone of
\fBpackage\-1\fR (a multi-socket system).
.SH "REMARKS"
.LP
Administrative (root) privileges are usually needed to use
\fBrapl\-set\fR.
.LP
Setting constraint power cap and/or time limit values does not enable or
disable a zone/subzone - the \-e/-\-z\-enabled flag must be set
explicitly.
.LP
Power units: microwatts (uW)
.br
Time units: microseconds (us)
.SH "BUGS"
.LP
The following are behavioral quirks due to the kernel interface or
abnormalities in some hardware.
They are not bugs in \fBrapl\-set\fR and should not be reported as such.
.LP
The kernel is not guaranteed to accept power limit and time window values
exactly as specified.
Values may be rounded due to how they are encoded in Model-Specific
Registers.
This rounding error is usually small within normal operating ranges, but
can be significant at the extremes.
.LP
Some systems may not accept time window values correctly, requiring a
system reboot to reset the time window to the default value.
.LP
Values returned by the kernel, e.g., by rapl\-info(1), sometimes lose
accuracy from the actual stored value due to integer rounding.
For example, the kernel may return a time window value of 7812 us, when in
fact the actual stored value is 7812.5 us.
Writing back an integer value returned by the kernel may cause unexpected
changes to its stored value due to these rounding problems.
.LP
At the time of this writing, some features in the powercap interface are not
supported by RAPL.
Resetting a zone energy counter (\-j/\-\-z\-energy) may result in an EACCES
(Permission denied) error due to the energy_uj file being read-only.
.LP
Some systems may not allow disabling package-level zones/constraints.
.LP
It is possible for packages on multi-socket systems to be indexed out of
order by the kernel.
For example, the package at index 0 (\fB\-p 0\fR) could actually be named
\fBpackage\-1\fR while the package at index 1 (\fB\-p 1\fR) is named
\fBpackage\-0\fR.
If this matters to the user, it is their responsibility to check before
setting values, e.g., using rapl\-info(1).
.LP
Report bugs upstream at <https://github.com/powercap/powercap>
.SH "FILES"
.nf
\fI/sys/devices/virtual/powercap/intel\-rapl/*\fP
.nf
\fI/sys/class/powercap/intel\-rapl/*\fP
.fi
.SH "AUTHORS"
.nf
Connor Imes <connor.k.imes@gmail.com>
.fi
.SH "SEE ALSO"
.BR powercap\-info (1),
.BR powercap\-set (1),
.BR rapl\-info (1)

RELEASES.md
# Release Notes

## [v0.4.0] - 2021-04-25

### Fixed

* powercap.h: header documentation for 'powercap_*_file_get_name' functions' return behavior
* rapl-{info,set}: deprecation messages that specified non-existent alternatives

### Added

* Functions to support opening files in powercap.h:
  * 'powercap_control_type_file_open'
  * 'powercap_zone_file_open'
  * 'powercap_constraint_file_open'
* CMake helper to find powercap without relying on pkg-config

### Changed

* Binaries powercap-{info,set} now accept the control type argument without requiring a preceding -p flag
* Increased minimum CMake version from 2.8.12 to 3.1

### Deprecated

* Flag '-p/--control-type' in powercap-{info,set} - specify control type name as the first positional argument instead
* Add compiler deprecation attributes to deprecated functions


## [v0.3.1] - 2020-11-08

### Fixed

* [#6] powercap-common-test:test_snprintf_base_path fails on ppc64el architecture


## [v0.3.0] - 2020-08-15

### Added

* Support for top-level control type
  * Struct 'powercap_control_type', enum 'powercap_control_type_file', and associated functions in powercap.h:
    * 'powercap_control_type_get_enabled'
    * 'powercap_control_type_set_enabled'
    * 'powercap_control_type_file_get_name'
  * Functions in powercap-sysfs.h:
    * 'powercap_sysfs_control_type_get_enabled'
    * 'powercap_sysfs_control_type_set_enabled'
  * Functions in powercap-rapl.h:
    * 'powercap_rapl_control_is_supported'
    * 'powercap_rapl_control_is_enabled'
    * 'powercap_rapl_control_set_enabled'
  * Argument '--enabled' to powercap-{info,set} for getting/setting control type enabled field
* Function 'powercap_rapl_get_num_instances' in powercap-rapl.h (supersedes 'powercap_rapl_get_num_packages')
* Argument '--nconstraints' to powercap-info for getting the number of zone constraints

### Changed

* Increased minimum CMake version from 2.8.5 to 2.8.12 to support target_compile_definitions
* Updated 'powercap-{info,set}' man pages
* Argument '--zone' for powercap-set no longer required due to introduction of '--enabled' argument
* Disabled logging by default (in general, libraries shouldn't print output unless requested)

### Deprecated

* Binaries rapl-{info,set} - use powercap-{info,set} instead
* Interface powercap-rapl-sysfs.h - use powercap-sysfs.h directly instead
* Function 'powercap_rapl_get_num_packages' in powercap-rapl.h - use 'powercap_rapl_get_num_instances' instead

### Fixed

* Fixed [#4]: powercap-rapl now checks if parent zone is PACKAGE or PSYS


## [v0.2.0] - 2019-12-03

### Added

* This RELEASES.md file
* Multiarch support (use GNU standard installation directories)
* Additional documentation in README
* Function 'rapl_sysfs_zone_exists' in powercap-rapl-sysfs.h
* Command line long option '--zone' to rapl-info and rapl-set
* Functions to powercap-rapl-sysfs.h allowed by the powercap API that RAPL doesn't support (but could in the future)
* Options to rapl-info and rapl-set to support optional powercap features not currently implemented in RAPL

### Changed

* Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs
* More pedantic man page source formatting
* RAPL sysfs 'pkg' parameter names changed to 'zone' to improve generality
* RAPL 'package' parameter names changed to 'id' to improve generality
* On failure, binaries now exit with positive error codes instead of negative values

### Deprecated

* Function 'rapl_sysfs_pkg_exists' in powercap-rapl-sysfs.h - use 'rapl_sysfs_zone_exists' instead
* Function 'rapl_sysfs_sz_exists' in powercap-rapl-sysfs.h - use 'rapl_sysfs_zone_exists' instead
* Command line long option '--package' for rapl-info and rapl-set - use '--zone' instead

### Removed

* Removed private symbol exports in shared object library (already patched in Debian)
* Undocumented (and unintended) --package long option from powercap-set

### Fixed

* Fixed [#1]: Kernel uses hexadecimal numbers in directory paths, not decimal


## [v0.1.1] - 2017-09-21

### Added

* Added stateless interfaces to sysfs
* Added powercap-info, powercap-set, rapl-info, and rapl-set binaries and man pages
* Added VERSION and SOVERSION to shared object libraries

### Changed

* Update license to use author as copyright holder


## v0.1.0 - 2017-06-09

### Added

* Initial public release

[v0.4.0]: https://github.com/powercap/powercap/compare/v0.3.1...v0.4.0
[v0.3.1]: https://github.com/powercap/powercap/compare/v0.3.0...v0.3.1
[v0.3.0]: https://github.com/powercap/powercap/compare/v0.2.0...v0.3.0
[v0.2.0]: https://github.com/powercap/powercap/compare/v0.1.1...v0.2.0
[v0.1.1]: https://github.com/powercap/powercap/compare/v0.1.0...v0.1.1
[#6]: https://github.com/powercap/powercap/issues/6
[#4]: https://github.com/powercap/powercap/issues/4
[#1]: https://github.com/powercap/powercap/issues/1


utils/rapl-set.c
/*
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Set RAPL values.
 *
 * @author Connor Imes
 * @date 2017-08-24
 */
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include "powercap-rapl-sysfs.h"
#include "util-common.h"

static const char short_options[] = "hp:z:c:je:l:s:";
static const struct option long_options[] = {
  {"help",                no_argument,        NULL, 'h'},
  {"package",             required_argument,  NULL, 'p'},
  {"zone",                required_argument,  NULL, 'p'},
  {"subzone",             required_argument,  NULL, 'z'},
  {"constraint",          required_argument,  NULL, 'c'},
  {"z-energy",            no_argument      ,  NULL, 'j'},
  {"z-enabled",           required_argument,  NULL, 'e'},
  {"c-power-limit",       required_argument,  NULL, 'l'},
  {"c-time-window",       required_argument,  NULL, 's'},
  {0, 0, 0, 0}
};

static void print_usage(void) {
  printf("\nThis utility is deprecated, use powercap-set instead.\n\n");
  printf("Usage: rapl-set [OPTION]...\n");
  printf("Options:\n");
  printf("  -h, --help                   Print this message and exit\n");
  printf("  -p, --zone=ZONE              The zone number (0 by default)\n");
  printf("      --package=PACKAGE        Deprecated, use --zone instead\n");
  printf("  -z, --subzone=SUBZONE        The subzone number (none by default)\n");
  printf("  -c, --constraint=CONSTRAINT  The constraint number (none by default)\n");
  printf("The following is a zone-level argument (-z/--subzone is optional):\n");
  printf("  -j, --z-energy               Reset zone energy counter\n");
  printf("  -e, --z-enabled=1|0          Enable/disable a zone\n");
  printf("The following constraint-level arguments may be used together and require -c/--constraint (-z/--subzone is optional):\n");
  printf("  -l, --c-power-limit=UW       Set constraint power limit\n");
  printf("  -s, --c-time-window=US       Set constraint time window\n");
  printf("\nPower units: microwatts (uW)\n");
  printf("Time units: microseconds (us)\n");
}

static void print_common_help(void) {
  printf("Considerations for common errors:\n");
  printf("- Ensure that the intel_rapl kernel module is loaded\n");
  printf("- Ensure that you run with administrative (super-user) privileges\n");
  printf("- Resetting a zone energy counter is an optional powercap feature and may not be supported by RAPL\n");
}

int main(int argc, char** argv) {
  u32_param zone = {0, 0};
  u32_param subzone = {0, 0};
  u32_param constraint = {0, 0};
  int reset_energy = 0;
  u32_param enabled = {0, 0};
  u64_param power_limit = {0, 0};
  u64_param time_window = {0, 0};
  int c;
  int cont = 1;
  int ret = 0;

  /* Parse command-line arguments */
  while (cont) {
    c = getopt_long(argc, argv, short_options, long_options, NULL);
    switch (c) {
    case -1:
      cont = 0;
      break;
    case 'h':
      print_usage();
      return EXIT_SUCCESS;
    case 'p':
      ret = set_u32_param(&zone, optarg, &cont);
      break;
    case 'z':
      ret = set_u32_param(&subzone, optarg, &cont);
      break;
    case 'c':
      ret = set_u32_param(&constraint, optarg, &cont);
      break;
    case 'j':
      if (reset_energy) {
        cont = 0;
        ret = -EINVAL;
      }
      reset_energy = 1;
      break;
    case 'e':
      ret = set_u32_param(&enabled, optarg, &cont);
      break;
    case 'l':
      ret = set_u64_param(&power_limit, optarg, &cont);
      break;
    case 's':
      ret = set_u64_param(&time_window, optarg, &cont);
      break;
    case '?':
    default:
      cont = 0;
      ret = -EINVAL;
      break;
    }
  }

  /* Verify argument combinations */
  if (ret) {
    fprintf(stderr, "Invalid arguments\n");
  } else if (constraint.set && !(power_limit.set || time_window.set)) {
    fprintf(stderr, "Must set at least one constraint-level argument when using -c/--constraint\n");
    ret = -EINVAL;
  } else if (!constraint.set && (power_limit.set || time_window.set)) {
    fprintf(stderr, "Must specify -c/--constraint when using constraint-level arguments\n");
    ret = -EINVAL;
  } else if (!(reset_energy || enabled.set || power_limit.set || time_window.set)) {
    printf("Nothing to do\n");
    ret = -EINVAL;
  }
  if (ret) {
    print_usage();
    return EXIT_FAILURE;
  }

  /* Check if zone/subzone/constraint exist */
  if (rapl_sysfs_zone_exists(zone.val, 0, 0)) {
    fprintf(stderr, "Zone does not exist\n");
    ret = -EINVAL;
  } else if (subzone.set && rapl_sysfs_zone_exists(zone.val, subzone.val, 1)) {
    fprintf(stderr, "Subzone does not exist\n");
    ret = -EINVAL;
  } else if (constraint.set && rapl_sysfs_constraint_exists(zone.val, subzone.val, subzone.set, constraint.val)) {
    fprintf(stderr, "Constraint does not exist\n");
    ret = -EINVAL;
  }
  if (ret) {
    print_common_help();
    return EXIT_FAILURE;
  }

  /* Perform requested action(s) */
  if (reset_energy) {
    c = rapl_sysfs_zone_reset_energy_uj(zone.val, subzone.val, subzone.set);
    if (c) {
      perror("Error setting energy counter");
      ret |= c;
    }
  }
  if (enabled.set) {
    c = rapl_sysfs_zone_set_enabled(zone.val, subzone.val, subzone.set, enabled.val);
    if (c) {
      perror("Error setting enabled/disabled");
      ret |= c;
    }
  }
  if (power_limit.set) {
    c = rapl_sysfs_constraint_set_power_limit_uw(zone.val, subzone.val, subzone.set, constraint.val, power_limit.val);
    if (c) {
      perror("Error setting power limit");
      ret |= c;
    }
  }
  if (time_window.set) {
    c = rapl_sysfs_constraint_set_time_window_us(zone.val, subzone.val, subzone.set, constraint.val, time_window.val);
    if (c) {
      perror("Error setting time window");
      ret |= c;
    }
  }
  if (ret) {
    print_common_help();
  }

  return ret ? EXIT_FAILURE : EXIT_SUCCESS;
}
